File: 0_8_detect_cycle_in_directed-graph.py
import threading
# directed graph
class Graph:
    def __init__(self, vertex):
        # vertex of the graph
        self.vertex = vertex

        # adjacent list
        self.adjList = [[] for x in range(vertex)]

        # visited
        self.visit = [False for x in range(vertex)]

        # stack to check for the cycle
        self.stack = [False for x in range(vertex)]

        # mark when going through checking whether the given graph having cycle or not
        self.cycle = False

    # explore a particular vertex
    def Explore(self, vertex):
        # mark the vertex as visited
        self.visit[vertex]=True

        # make the stack of this vertex as true
        self.stack[vertex]=True

        # now explore the neighbours of the vertex
        for ver in  self.adjList[vertex]:
            # if not visited then explore it
            if self.visit[ver]==False:
                self.Explore(ver)
            # if not visited and then it is in the stack then it makes  the cycle
            elif self.stack[ver]==True:
                self.cycle=True
                return 
            else:
                pass

        # now if we have explored then we need to mark down from True to False of that vertex in te self.stack
        self.stack[vertex]=False


    # dfs
    def dfs(self):
        for vertex in range(self.vertex):
            # check if not visisted that is explored then explore that particualr vertex
            if self.visit[vertex]==False:
                self.Explore(vertex)


    # add edges
    def addEdge(self, vertex1, vertex2):
        self.adjList[vertex1].append(vertex2)

def main():
    # number of vertices and edges
    vertex, edge = [int(x) for x in input().split()]

    # graph 
    graph = Graph(vertex)

    # assign the edges of the graph
    for x in range(edge):
        vertex1, vertex2 = [int(y) for y in input().split()]
        # add this edge
        graph.addEdge(vertex1, vertex2)

    # do dfs to detect the cycle
    graph.dfs()
    
    print("Graph is having the cycle") if graph.cycle==True else print("Graph do not contains the cycle")

threading .Thread(target=main).start()

    
'''
tets case
input
4 6
0 1
0 2
1 2
2 0
2 3
3 3

output
Graph is having the cycle
'''

File: 0_8_detect_cycle_in_directed-graph.java
import java.util.*;
public class FindCycle {
    // number of vertices
    private int vertices;
    // adjacency list
    private List<Integer> adjList[];
    // for visit
    private boolean visit[];
    // array for checking 
    private boolean check[];
    // cycle
    private boolean cycle=false;

    FindCycle(int vertex){
        vertices = vertex;
        // adjacency list 
        adjList = new ArrayList[vertex];
        // check 
        check = new boolean[vertex];
        // visit
        visit = new boolean[vertex];

        for (int x=0; x<vertex; x++){
            //  assigning new list to the at each index
            adjList[x] = new ArrayList();

            visit[x]=false;
            check[x]=false;
        }
    }

    // add Edge
    void addEdge(int vertex1, int vertex2){
        adjList[vertex1].add(vertex2);
    }

    // explore
    void Explore(int vertex){
        // first mark as the visited
        visit[vertex]=true;

        // make in the check as true 
        check[vertex]=true;

        // now iterate through it's neighbour
        Iterator<Integer> i = adjList[vertex].listIterator();

        while(i.hasNext()){
            int ver = i.next();
            if (visit[ver]==false){
                Explore(ver);
            }
            else if(check[ver]==true){
                cycle=true;
                return;
            }
            else{
                ;
            }
        }
    }

    // dfs
    void dfs(){
        //  go through all the edges
        for (int vertex=0; vertex<vertices; vertex++){
            //  if not visited then explotre the vertex
            if (visit[vertex]==false){
                Explore(vertex);
            }
        }
    }
    
    public static void main(String[] args){
        Scanner scanner = new Scanner(System.in);

        int vertex = scanner.nextInt();
        int edge = scanner.nextInt();

        FindCycle graph=new FindCycle(vertex);
        //  add the edges
        for(int x=0; x<edge; x++){
            int vertex1 = scanner.nextInt();
            int vertex2 = scanner.nextInt();

            //  add edge
            graph.addEdge(vertex1, vertex2);
        }

        graph.dfs();

        if (graph.cycle==true) System.out.println("Graph is having the cycle");
        else System.out.println("Graph do not contains the cycle");

    }
}
